packages/core/src/utils/JsonTransformer.ts

```diff
@@ -1,15 +1,53 @@
+import type { Validate } from 'class-validator'
+
 import { instanceToPlain, plainToInstance, instanceToInstance } from 'class-transformer'
 
+import { ClassValidationError } from '../error/ClassValidationError'
+import { isValidationErrorArray } from '../error/ValidationErrorUtils'
+
+import { MessageValidator } from './MessageValidator'
+
+interface Validate {
+  validate?: boolean
+}
+
 export class JsonTransformer {
   public static toJSON<T>(classInstance: T) {
     return instanceToPlain(classInstance, {
       exposeDefaultValues: true,
     })
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  public static fromJSON<T>(json: any, Class: { new (...args: any[]): T }): T {
-    return plainToInstance(Class, json, { exposeDefaultValues: true })
+  public static fromJSON<T>(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    json: any,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    Class: { new (...args: any[]): T },
+    options: Validate = { validate: true }
+  ): T {
+    if (!options.validate) {
+      return plainToInstance(Class, json, { exposeDefaultValues: true })
+    } else {
+      const plainInstance = plainToInstance(Class, json, { exposeDefaultValues: true })
+      //  validateSync is not happy with null/undefined. Return it to keep returning the same as previous versions without validation
+      if (plainInstance === undefined || plainInstance === null) {
+        return plainInstance
+      }
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const errors = MessageValidator.validateSync(plainInstance as any)
+      // NOTE: validateSync (strangely) return an Array of errors so we
+      // have to transform that into an error of choice and throw that.
+      if (isValidationErrorArray(errors)) {
+        throw new ClassValidationError('Failed to validate class.', {
+          recordType: typeof Class,
+          cause: new Error(JSON.stringify(errors)),
+        })
+      } else if (errors.length !== 0) {
+        throw new ClassValidationError('An unknown validation error occurred.', { recordType: typeof Class })
+      } else {
+        return plainInstance as T
+      }
```

Comment:
Nit: I like to keep happy flow top level:

```suggestion

     return plainInstance as T
```